docs/gni_sources.md

```diff
@@ -1,3 +1,84 @@
-Use of sources.gni should be limited to sources that would otherwise have circular dependencies that cannot be easily resolved, in other cases files should be added a new or existing BUILD.gn targets. The chromium ios code is a good model for separating out dependencies and sometimes makes use of interface/implementation patterns. A good example in desktop/android code is //chrome/browser/profiles target [https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/profiles/BUILD.gn;l=8](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/profiles/BUILD.gn;l=8) which has a minimal set of dependencies and the bulk of the implementation that would likely trigger a lot of circular deps is in profile_impl.cc in the //chrome/browser target. This reduces the number of sources/targets that either depend on //chrome/browser or have to be added to //chrome/browser sources list. Other strategies used to break //chrome dependencies in components can also sometimes be applied to //brave/browser code to avoid circular dependencies [https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion](https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion). Do not use `check_includes = false`.
+### sources.gni
 
-This does not mean that you cannot use sources.gni to inject sources into existing targets, but please try to avoid it when practical.
+#### Background
+The use of sources.gni was initially added to help deal with the issue of
+circular dependencies. We often subclass upstream code like
+`BraveContentBrowserClient` -> `ChromeContentBrowserClient`, but in a separate
+target there would be a circular dependency because `BraveContentBrowserClient`
+depends on `//chrome/brave` and `//chrome/brave` would have to depend on the
+target that contains `BraveContentBrowserClient`. This actually originated from
+gyp which had very large targets that upstream has been slowly working to split
+up. This is mainly an issue for `//chrome/browser` and `//chrome/browser/ui` and
+as upstream is working to break these targets up, we need to make sure we are
+not adding to the problem on our side.
+
+#### Usage of sources.gni
+
+Use of sources.gni to include sources in `//chrome/browser` and
+`//chrome/browser/ui` should be avoided. Adding deps through sources.gni is
```

Comment:
do we have a good example where this statement holds and it allows us to actually use circular deps without any issues?

Rolling this doc without a working example doesn't seem right to me.

I'd rather have a handful of successful migration cases from `sources.gni` to circular deps before going all in and making this a rule.

Reply:
> do we have a good example where this statement holds and it allows us to actually use circular deps without any issues?

Well, there are always potential issues which is why the upstream targets have the public dependencies targets, but these are the two main examples where they are doing this and we have been doing it for a while now as well and have many successful migration cases.

Reply:
also this is not a new rule, this has been mentioned in slack before and @brave/sources-gni-reviewers are supposed to be enforcing this